### ISSUE TITLE
Can't use ktor-ktor-client-core with webpack

### ISSUE THREAD

**Issue by gagarin55** [ktorio/ktor#1940]
Webpack can't find dynamic require which is located in the following code https://github.com/ktorio/ktor/blob/916cff4b964573cf5f7cac30eb42c656890902ad/ktor-client/ktor-client-core/js/src/io/ktor/client/engine/js/compatibility/Utils.kt#L58

Webpack bundled code says *Error load 'node-fetch' module*

**Comment by toth-istvan-zoltan** [comment]
This stopped me dead on Saturday, my solution for the time being. I know it is not right, but it lets me go on with development.

I have a multiplatform project with browser target only (probably works with nodejs target too).

1. created a `patch` directory in my project root
2. copied `build/js/packages_imported/ktor-ktor-client-core/1.3.2/ktor-ktor-client-core.js` into `patch`
3. edited `ktor-ktor-client-core.js` patch manually, I simply removed jsRequire and then **ONE** place its called from
4. added a task to my `build.gradle.kts` to copy after the `compileKotlinJs` task

```
val pathKtorIoClient by tasks.registering(Copy::class) {
    from("$projectDir/patch/")
    into("$projectDir/build/js/packages_imported/ktor-ktor-client-core/1.3.2")
    include("*.js")
}

tasks["compileKotlinJs"].finalizedBy(pathKtorIoClient)
```

**Comment by e5l** [comment]
Hi @gagarin55, could you share the details of your setup: what is your platform(node.js or browser)?
Could you also list ktor related Gradle dependencies?

**Comment by gagarin55** [comment]
> Hi @gagarin55, could you share the details of your setup: what is your platform(node.js or browser)?
> Could you also list ktor related Gradle dependencies?

Hello, @e5l 
Nothing special, just kotlin/js nodejs package, but this package can't be used in webpacked projects.

Part of build.gradle.kts file:
```
...
val ktor_version = "1.3.2"
val serialization_version = "0.20.0"

repositories {
    jcenter()
    maven("https://kotlin.bintray.com/kotlinx")
    maven("https://dl.bintray.com/kotlin/ktor")
    mavenCentral()
}

dependencies {
    implementation(kotlin("stdlib-js"))

    // ----- Ktor Framework
    implementation("io.ktor:ktor-client-js:$ktor_version")
    implementation("io.ktor:ktor-client-json-js:$ktor_version")
    implementation("io.ktor:ktor-client-serialization-js:$ktor_version")
    implementation("org.jetbrains.kotlinx:kotlinx-serialization-runtime-js:$serialization_version")

    // Due to issue https://github.com/ktorio/ktor/issues/1822
    // we have to import transitive npm deps

    implementation(npm("abort-controller"))
    implementation(npm("node-fetch"))
    implementation(npm("text-encoding"))
    // ------ End of Ktor Framework

    testImplementation("org.jetbrains.kotlin:kotlin-test-js")

}

kotlin.target.nodejs {
}

kotlin {
    target {
        useCommonJs()
    }
}

```

**Comment by ScottPierce** [comment]
I'm encountering this same error:
```
Error loading module 'node-fetch': Error: Cannot find module 'node-fetch'
```

I'm using webpack as well.

**Comment by toth-istvan-zoltan** [comment]
Hi, check this: https://github.com/ktorio/ktor/issues/961#issuecomment-650695064

**Comment by ScottPierce** [comment]
@toth-istvan-zoltan The error I'm getting is at runtime, not at webpack time. If I'm understanding this right, telling webpack that those files are there at runtime won't solve this, because this is already failing at runtime.

**Comment by toth-istvan-zoltan** [comment]
Are you running it in node or in browser? If you running it in a browser then you do not have node-fetch and should use the browser fetch. I think :) I'm not really familiar with node to be honest.

**Comment by ScottPierce** [comment]
I'm running it in node.

**Comment by toth-istvan-zoltan** [comment]
@ScottPierce Hm... assuming you have the import as in the comment of gagarin55's I have no idea what goes wrong.

Anyway, please let me know if you find a solution, this is a problematic point for me as well. Right now I use a manually patched JS source. There is a change commit in master related to this, but I haven't got the time to check it out yet.

**Comment by ScottPierce** [comment]
I tried the exclude. This is still showing up in the webpack code, and causing an error though:
```
  if (util.PlatformUtils.IS_BROWSER) {
    tmp$ = fetch(closure$input, closure$init);
  } else {
    tmp$ = jsRequire('node-fetch')(closure$input, closure$init);
  }
```

**Comment by toth-istvan-zoltan** [comment]
@ScottPierce If the webpack error is the critical dependency, then right now I don't know any other way than patch ktor-ktor-client-core.js in build/js/packages_imported.

I've added a small gradle task that copies the patched version of the file into the directory. This is plain **wrong** but it kept me going. A tad better would be to use a small patch, but I haven't figured out how to do that with gradle yet.

```kotlin
val pathKtorIoClient by tasks.registering(Copy::class) {
    from("${rootProject.projectDir}/comm/patch/")
    into("${rootProject.projectDir}/build/js/packages_imported/ktor-ktor-client-core/1.3.2")
    include("*.js")
}

tasks["compileKotlinJs"].finalizedBy(pathKtorIoClient)
```

**Comment by ScottPierce** [comment]
@toth-istvan-zoltan how should the file be patched? I'm open to doing this, but I'm not sure what I should change it to.

**Comment by toth-istvan-zoltan** [comment]
@ScottPierce Well, I went to the line in question, 16801 if I remember correctly, and simply changed the require(module) line to require('node-fetch').

The point is that in the latest stable version the jsRequire is called only once and only with the static paramter 'node-fetch'. So the change is safe.

It is important to copy the file into a dir (comm/pacth for me), do the change there and add the gradle copy task, otherwise new builds / gradle clean will remove the patched file.

I've submitted a patch and it's been already merged into master but it haven't reached stable release yet.

As the GitHub build currently fails you can't use the latest snapshot I think.

I've tried to make a new build for myself this morning but there's been a Mac OS upgrade with moved curl and then I ran another issue with atomicfu c interop, so I've given up.

**Comment by gagarin55** [comment]
I've found a solution - don't use Kotlin/JS anymore at all. This issue is one among many flaws of this tech.

**Comment by oleg-larshin** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/KTOR-344">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.

**Comment by e5l** [comment]
Can't reproduce since 1.4.0
